### PR TITLE
Jms Service Migration with shared FIle/JDBC store in MII domain 

### PR DESCRIPTION
**Pull request by anpanigr**
This test class verifies JMS Service migration when a cluster is scaled  down by removing the pod. 
  The usecase verifies the JMS service on shutdown  pod/server is migrated to one of the active server and the JMS 
   messages can be recovered after service migration.
   The dynamic cluster is configured with an ORACLE leasing datasource.

 The cluster targeted persistent store is configured with 'Dynamic'  distribution policy and 'Always' migration policy.
 The associated file store directory is on a shared volume so that it can  be accessed from each managed server pod/server.
  The associated JDBC store  is on remote DB instance so that it can  be accessed from each managed server pod/server.
 
An uniform distributed queue is configured with a cluster targeted JMS  system resource with one member on each managed server.
  (a) Test client sends 100 messages to member queue@managed-server2
  (b) Scale down the cluster with replica count 1
  (c) Make sure all 100 messages got recovered once the 
      JMS Service@managed-server2 is migrated to managed-server1 

Remote Jenkin URL :  
  https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4284/ ( all test passed)
  https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4305/   ( known intermittent failure) 
  https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4316/   ( known  failure in MiiSample)
  https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4332/
  https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4336/

### CODE REVIEW

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+    ecmd.append(cpUrl);
+    ecmd.append(" oracle.jdbc.OracleDriver");
+    ecmd.append(" -verbose ");
+    ecmd.append(" -u \"sys as sysdba\"");
+    ecmd.append(" -p Oradoc_db1");
+    ecmd.append(" /u01/oracle/leasing.ddl");
+    ExecResult execResult = assertDoesNotThrow(
+        () -> execCommand(domainNamespace, adminServerPodName,
+            null, true, "/bin/sh", "-c", ecmd.toString()));
+    assertTrue(execResult.exitValue() == 0, "Could not poulate the Leasing Table");
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:354]:
```suggestion
    assertTrue(execResult.exitValue() == 0, "Could not create the Leasing Table");
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+  /**
+   * Verify JMS Service is migrated to an available active server.
+   * Here the JMS messages are stored in File store on PV
+   */
+  @Test
+  @Order(1)
+  @DisplayName("Verify JMS Service migration with FileStore")
+  public void testMiiJmsServiceMigrationWithFileStore() {
+
+    buildClientOnPod();
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:280]: call CommonTestUtils.runJavacInsidePod. see ItMiiDynamicUpdate.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+    runJmsClientOnAdminPod("send",
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:281]: see if you can call CommonTestUtils.runClientInsidePod.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+    logger.info("checkManagedServerConfiguration: curl command returned {0}", result.toString());
+    if (result.stdout().equals("200")) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // Crate a ConfigMap with a model file to add a new WebLogic cluster
+  private void createClusterConfigMap(String configMapName, String modelFile) {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:610]: looks like this is not used, can be removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+            String.format("podDoesNotExist failed with ApiException for %s in namespace in %s",
+                podName, domNamespace)));
+  }
+
+  /*
+   * Verify the server MBEAN configuration through rest API.
+   * @param managedServer name of the managed server
+   * @returns true if MBEAN is found otherwise false
+   **/
+  private boolean checkManagedServerConfiguration(String managedServer) {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:582]: looks like this is not used, can be removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+    logger.info("Create domain custom resource for domainUid {0} in namespace {1}",
+            domainUid, domNamespace);
+    boolean domCreated = assertDoesNotThrow(() -> createDomainCustomResource(domain),
+            String.format("Create domain custom resource failed with ApiException for %s in namespace %s",
+                    domainUid, domNamespace));
+    assertTrue(domCreated, String.format("Create domain custom resource failed with ApiException "
+                    + "for %s in namespace %s", domainUid, domNamespace));
+  }
+
+  private void checkPodNotCreated(String podName, String domainUid, String domNamespace) {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:563]: looks like this is not used, can be removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+    V1ConfigMap configMap = new V1ConfigMap()
+        .data(data)
+        .metadata(meta);
+
+    boolean cmCreated = assertDoesNotThrow(() -> createConfigMap(configMap),
+        String.format("Can't create ConfigMap %s", configMapName));
+    assertTrue(cmCreated, String.format("createConfigMap failed while creating ConfigMap %s", configMapName));
+  }
+
+  private ExecResult checkJdbcRuntime(String resourcesName) {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:634]: looks like this is not used, can be removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+          String podLog = assertDoesNotThrow(() -> getPodLog(pods.get(0).getMetadata().getName(), namespace),
+              "Failed to get pod log");
+          logger.severe(podLog);
+          fail("Change permissions on PV hostpath job failed");
+        }
+      }
+    }
+  }
+
+  private void checkLogsOnPV(String commandToExecuteInsidePod, String podName) {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:705]: looks like this is not used, can be removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 9]
```diff
@@ -0,0 +1,739 @@
...
+    try {
+      result = exec(new String(curlString), true);
+    } catch (Exception ex) {
+      logger.info("checkJdbcRuntime: caught unexpected exception {0}", ex);
+      return null;
+    }
+    return result;
+  }
+
+  private static void createJobToChangePermissionsOnPvHostPath(String pvName, String pvcName, String namespace) {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:655]: this method is in CommonMiiTestUtils, call that

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 10]
```diff
@@ -0,0 +1,739 @@
...
+    boolean secretCreated = assertDoesNotThrow(() -> createSecret(new V1Secret()
+            .metadata(new V1ObjectMeta()
+                    .name(secretName)
+                    .namespace(domNamespace))
+            .stringData(secretMap)), "Create secret failed with ApiException");
+    assertTrue(secretCreated, String.format("create secret failed for %s in namespace %s", secretName, domNamespace));
+
+  }
+
+  private static void createDomainSecret(String secretName, String username, String password, String domNamespace)
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:481]: check CommonMiiTestUtils for this method

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 11]
```diff
@@ -0,0 +1,739 @@
...
+    assertTrue(psuccess3,
+        String.format("Cluster replica patching failed for domain %s in namespace %s", domainUid, domainNamespace));
+    checkPodDoesNotExist(managedServerPrefix + "2", domainUid, domainNamespace);
+
+    runJmsClientOnAdminPod("receive", 
+            "JdbcJmsServer@managed-server2@jms.jdbcUniformQueue");
+  }
+
+  // Create leasing Table (ACTIVE)
+  private static void createLeasingTable() {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:322]: add this to DBUtils

**Review by vanajamukkara — CHANGES_REQUESTED** [review]
There is lot of duplicate and unused code which needs to be removed/modified.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 12]
```diff
@@ -0,0 +1,739 @@
...
+    // check managed server services and pods are ready
+    for (int i = 1; i <= replicaCount; i++) {
+      logger.info("Wait for managed server services and pods are created in namespace {0}",
+          domainNamespace);
+      checkPodReadyAndServiceExists(managedServerPrefix + i, domainUid, domainNamespace);
+    }
+  }
+
+  /**
+   * Verify all server pods are running.
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:263]: removed the empty method BeforeEach()

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 13]
```diff
@@ -0,0 +1,739 @@
...
+  private static int dbNodePort;
+
+  private final Path samplePath = Paths.get(ITTESTS_DIR, "../kubernetes/samples");
+  private final Path domainLifecycleSamplePath = Paths.get(samplePath + "/scripts/domain-lifecycle");
+
+  /**
+   * Install Operator.
+   * Create domain resource defintion.
+   * @param namespaces list of namespaces created by the IntegrationTestWatcher by the
+   JUnit engine parameter resolution mechanism
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:160]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 1 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:354]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 7 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:634]: removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 8 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:705]: removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 10 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:481]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 9 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:655]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 3 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:281]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 2 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:280]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 11 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:322]: This method is specific to Service Migration usecase not a common utility to create any generic table.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+
+/**
+ * This test class verifies JMS Service migration when a cluster is scaled 
+ * down by removing the pod. The usecase verifies the JMS service on shutdown 
+ * pod/server is migrated to one of the active server and the JMS messages
+ * can be recovered after service migration.
+ * The dynamic cluster is configured with an ORACLE leasing datasource.
+ * The cluster targeted persistent store is configured with 'Dynamic' 
+ * distribution policy and 'Always' migration policy.
+ * The associated file store directory is on a shared volume so that it can
```
Comment (sankarpn) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:102]: From the description its not clear the test class verifies JMS service migration from 2 different persistent stores separately. Can you mention that it tests both stores and it is independent of each other.

**Review by sankarpn — CHANGES_REQUESTED** [review]

**Review by vanajamukkara — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 14 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java]: /**
 * This test class verifies JMS Service migration with JMS messages stored in 
 * (a) Persistent FileStore (b) Persistent JDBC Store accessible from all pods.
 * The dynamic cluster is scaled down to trigger the Service migration from 
 * a stopped pod/managed server to a live pod/managed server.
 * Configuration : 
 *   MII cluster domain with 2 managed servers
 *   Two set of JMS Resources with FileStore and JDBC Store
 *   All resources are targeted to cluster with enabled JMS service migration
 *   Two Distributed Queue(s) one with FileStore and the other with JDBC Store
 *   Separate ORACLE Datasource for cluster leasing
 * UseCase :  
 * (a) Test client sends 100 messages to member queue@managed-server2
 * (b) Scale down the cluster with replica count 1 to shutdown managed-server2
 * (c) Make sure the JMS Service@managed-server2 is migrated to managed-server1
 * (d) Make sure all 100 messages got recovered once the 
 *     JMS Service@managed-server2 is migrated to managed-server1 
 * */

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 4 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java]: removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 5 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java]: removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 6 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java]: removed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+    ExecResult execResult = assertDoesNotThrow(
+        () -> execCommand(domainNamespace, adminServerPodName,
+            null, true, "/bin/sh", "-c", ecmd.toString()));
+    assertTrue(execResult.exitValue() == 0, "Could not create the Leasing Table");
+  }
+
+  private void restartManagedServer(String serverName) {
+
+    String commonParameters = " -d " + domainUid + " -n " + domainNamespace;
+    CommandParams params;
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:360]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 16]
```diff
@@ -0,0 +1,516 @@
...
+    boolean result;
+    params = new CommandParams().defaults();
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:362]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 17]
```diff
@@ -0,0 +1,516 @@
...
+class ItMiiJmsRecovery {
+
+  private static String opNamespace = null;
+  private static String domainNamespace = null;
+  private static ConditionFactory withStandardRetryPolicy = null;
+  private static int replicaCount = 2;
+  private static final String domainUid = "mii-jms-recovery";
+  private static String pvName = domainUid + "-pv"; 
+  private static String pvcName = domainUid + "-pvc"; 
+  private StringBuffer curlString = null;
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:126]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 18]
```diff
@@ -0,0 +1,516 @@
...
+  private V1Patch patch = null;
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:127]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+  private static final String adminServerPodName = domainUid + "-admin-server";
+  private static final String managedServerPrefix = domainUid + "-managed-server";
+  private final String adminServerName = "admin-server";
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:130]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 20]
```diff
@@ -0,0 +1,516 @@
...
+  private final String clusterName = "cluster-1";
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:131]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 21]
```diff
@@ -0,0 +1,516 @@
...
+   * Get the JMSServer Runtime ClusterJmsServer@managed-server2 found on 
+   * specified managed server.
+   * @param managedServer name of managed server to look for JMSServerRuntime
+   * @returns true if MBEAN is found otherwise false
+   **/
+  private boolean checkJmsServerRuntime(String managedServer) {
+    ExecResult result = null;
+    int adminServiceNodePort
+        = getServiceNodePort(domainNamespace, getExternalServicePodName(adminServerPodName), "default");
+    StringBuffer curlString = new StringBuffer("status=$(curl --user weblogic:welcome1 ");
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:481]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 22]
```diff
@@ -0,0 +1,516 @@
...
+    boolean secretCreated = assertDoesNotThrow(() -> createSecret(new V1Secret()
+            .metadata(new V1ObjectMeta()
+                    .name(secretName)
+                    .namespace(domNamespace))
+            .stringData(secretMap)), "Create secret failed with ApiException");
+    assertTrue(secretCreated, String.format("create secret failed for %s in namespace %s", secretName, domNamespace));
+
+  }
+
+  private static void createDomainResource(
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java:401]: I need extra dbSecret parameter to create the domain resource.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java [hunk 23]
```diff
@@ -0,0 +1,516 @@
...
+        .conditionEvaluationListener(
+            condition -> logger.info("Wait for JMS Client to send/recv msg "
+                    + "(elapsed time {0}ms, remaining time {1}ms)",
+                condition.getElapsedTimeInMS(),
+                condition.getRemainingTimeInMS()))
+        .until(runClientInsidePod(adminServerPodName, domainNamespace,
+            "/u01", "JmsSendReceiveClient", "t3://" + domainUid + "-cluster-cluster-1:8001", action, queue, "100"));
+  }
+
+  private static void createDatabaseSecret(
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java]: I updated the createDomainResourceWithLogHome() in CommonMiiTestUtils.java to add a dataHome and used it in ItMiiJmsRecovery.java. 

Thanks, I could remove a big chunk of code from my class.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java (hunk 11 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiJmsRecovery.java]: Move the method to common/DBUtils.java

**Review by vanajamukkara — APPROVED** [review]
LG

**Pull request merged**